### PR TITLE
Remove the last of the $scope functions

### PR DESCRIPTION
**Pull request by garethbowen**
# Description

Refactors the remaining $scope functions affecting a wide range of code.

medic/cht-core#5791

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/e2e/send-message.js [hunk 1]
```diff
@@ -410,22 +412,7 @@ describe('Send message', () => {
...
-        const el = element(by.css('#message-header .name'));
-        return helper
-          .getTextFromElement(el)
-          .then(text => text === 'Unknown sender')
-          .catch(err => {
-            // item may have been removed from the page when the RHS is refreshed
-            console.log('Caught and ignoring an error trying to getText', err);
-          });
-      }, 12000);
+      clickLhsEntry(ALICE._id, 'Unknown sender');
```
Comment (garethbowen) [inline tests/e2e/send-message.js]: Reuse function in attempt to make this test more reliable. It didn't work (I fixed it elsewhere) but it's cleaner anyway so I left it in.

File: webapp/src/js/controllers/messages-content.js [hunk 2]
```diff
@@ -80,37 +81,32 @@ angular.module('inboxControllers').controller('MessagesContentCtrl',
...
-    var selectContact = function(id, type, options) {
-      options = options || {};
+    const markConversationReadIfNeeded = () => {
+      const hasUnreadDoc = ctrl.selectedMessage.messages.some(message => {
+        return !message.read && message.doc;
+      });
+      if (hasUnreadDoc) {
+        ctrl.markConversationRead();
+      }
+    };
```
Comment (garethbowen) [inline webapp/src/js/controllers/messages-content.js]: This makes the e2e test more stable (as well as the actual performance better) by only updating the redux state (and forcing an angular render) when the doc is not already read.

File: webapp/src/js/directives/messages-list.js [hunk 3]
```diff
@@ -18,6 +18,6 @@ angular.module('inboxDirectives').component('mmMessagesList', {
   controllerAs: 'messagesListCtrl',
   bindings: {
     loading: '<',
-    messages: '<'
+    conversations: '<'
```
Comment (garethbowen) [inline webapp/src/js/directives/messages-list.js]: I refactored this for clarity: "conversations" are groups of "messages" to/from a contact. Having both called "messages" was crazy...

File: webapp/src/js/services/send-message.js [hunk 4]
```diff
@@ -162,25 +163,25 @@ angular
...
           });
           self.setLastChangedDoc(doc);
           return doc;
         })
-        .then(function(doc) {
-          return DB().post(doc);
+        .then(doc => {
+          return $q.all([
+            DB().post(doc),
+            MarkRead([ doc ])
```
Comment (garethbowen) [inline webapp/src/js/services/send-message.js]: Another change to improve e2e test stability (as well as make the UI better) - if you're creating a doc it is automatically marked as read so it only needs to be rendered once and the user is not bugged to read a message they created.

File: webapp/tests/karma/unit/store/analytics.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    setupStore(initialState);
+    const selected = {};
+    analyticsActions.setSelectedAnalytics(selected);
+    const state = getState();
+    const analyticsState = selectors.getAnalyticsState(state);
+    chai.expect(state).to.not.equal(initialState);
+    chai.expect(analyticsState).to.deep.equal({ selected });
+  });
+
+});
```
Comment (garethbowen) [inline webapp/tests/karma/unit/store/analytics.js]: We used to have one huge test file for testing all the stores which meant it was really fragile. I broke it into ~6 separate files.

**Comment by garethbowen** [comment]
Review please @kennsippell . It's fairly huge, but most of it's moving code down from controllers to actions. Thanks for your time!

File: webapp/src/js/actions/messages.js [hunk 6]
```diff
@@ -42,13 +48,35 @@ angular.module('inboxServices').factory('MessagesActions',
         });
       }
 
+      function markConversationRead() {
```
Comment (kennsippell) [inline webapp/src/js/actions/messages.js]: nit: I would expect a function with this name to accept a conversation parameter or alike.... maybe `markSelectedConversationRead()`?

File: webapp/src/js/controllers/messages-content.js [hunk 7]
```diff
@@ -205,7 +186,7 @@ angular.module('inboxControllers').controller('MessagesContentCtrl',
               ctrl.allLoaded = conversation.length < 50;
               delete ctrl.firstUnread;
             }
-            markAllRead();
+            ctrl.markConversationRead();
```
Comment (kennsippell) [inline webapp/src/js/controllers/messages-content.js]: is this functionally equivalent? replacing `markAllRead()` with a call to `markSelectedConversationRead()` ?

File: webapp/src/js/controllers/messages.js [hunk 8]
```diff
@@ -37,103 +37,84 @@ angular
...
-        updateConversations({ changes: true });
-      },
-      filter: function(change) {
-        if (ctrl.currentTab !== 'messages') {
-          return false;
-        }
-        return (
+      callback: () => updateConversations({ merge: true }),
+      filter: change => {
+        return ctrl.currentTab === 'messages' &&
```
Comment (kennsippell) [inline webapp/src/js/controllers/messages.js]: is a set of parenthesis required here to maintain parity?

File: tests/e2e/send-message.js [hunk 9]
```diff
@@ -398,26 +386,4 @@ describe('Send message', () => {
     });
   });
 
-  describe('Display message without contact', () => {
```
Comment (kennsippell) [inline tests/e2e/send-message.js]: > i officially give up

lol

File: webapp/src/js/actions/contacts.js [hunk 10]
```diff
@@ -38,16 +134,93 @@ angular.module('inboxServices').factory('ContactsActions',
...
+                    canDelete: false, // this disables the "Delete" button in action bar until full load is complete
+                    canEdit: canEdit,
+                    childTypes: childTypes
+                  });
+                  return lazyLoadedContactData
+                    .then(() => {
+                      return $q.all([
+                        ContactSummary(selected.doc, selected.reports, selected.lineage),
+                        Settings(),
+                        getTasks()
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/actions/contacts.js]: Not only does it not return anything the next block isn't dependant on it finishing - it just registers a callback to update tasks when ready. I've renamed it and pulled it out of the `$q,all()` so it executes in its own time.

File: webapp/src/js/controllers/contacts-content.js [hunk 11]
```diff
@@ -85,20 +83,14 @@ angular.module('inboxControllers').controller('ContactsContentCtrl',
...
-          ctrl.unsetSelected();
-          $log.error('Error generating contact view model', err, err.message);
-        });
+      const getChildPlaces = !usersHomePlaceId || usersHomePlaceId !== id;
+      ctrl.setSelectedContact(id, { getChildPlaces }).catch(err => {
+        if (err.code === 404 && !silent) {
+          $translate('error.404.title').then(Snackbar);
+        }
+        ctrl.unsetSelected();
+        $log.error('Error generating contact view model', err, err.message);
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/controllers/contacts-content.js]: Updated.

File: webapp/src/js/controllers/contacts-report.js [hunk 12]
```diff
@@ -62,31 +64,36 @@ angular.module('inboxControllers').controller('ContactsReportCtrl',
       });
     };
 
-    var render = function(contact, options) {
-      $scope.setSelected(contact, options);
-      setCancelCallback();
-      return XmlForms.get($state.params.formId)
+    const render = function(id) {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/controllers/contacts-report.js]: Updated to take the contactId and the formId as params.

File: webapp/src/js/actions/global.js [hunk 13]
```diff
@@ -169,10 +171,90 @@ angular.module('inboxServices').factory('GlobalActions',
...
+      function openGuidedSetup() {
+        return Modal({
+          templateUrl: 'templates/modals/guided_setup.html',
+          controller: 'GuidedSetupModalCtrl',
+          controllerAs: 'guidedSetupModalCtrl',
+          size: 'lg',
+        }).catch(() => {}); // modal dismissed is ok
+      }
+
+      function deleteDoc(doc) {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/actions/global.js]: I feel the fact that there is a prompt is an implementation detail and not necessary to communicate in the function name. Other examples include the logout function which confirms before continuing, and the `$state.go` function which sometimes confirms before navigation.

File: webapp/src/js/directives/actionbar.js [hunk 14]
```diff
@@ -41,6 +46,23 @@ angular.module('inboxDirectives').directive('mmActionbar', function() {
       };
       const unsubscribe = $ngRedux.connect(mapStateToTarget, mapDispatchToTarget)(ctrl);
 
+      ctrl.bulkDelete = docs => {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/directives/actionbar.js]: I put as much in the directive/component/controller as possible. This one is only invoked from the action bar. The others are also invoked from other controllers. I'm not sure if this is the right approach - perhaps we need an actionbar actions as well...

File: webapp/src/js/actions/messages.js [hunk 15]
```diff
@@ -42,13 +48,35 @@ angular.module('inboxServices').factory('MessagesActions',
...
+        dispatch((dispatch, getState) => {
+          const state = getState();
+          const selected = Selectors.getSelectedMessage(state);
+          const conversations = Selectors.getConversations(state);
+          const docs = selected.messages.map(message => message.doc);
+          if (docs.length) {
+            const ids = docs.map(doc => doc._id);
+            const conversation = conversations.find(conversation => {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/actions/messages.js]: Somehow this comment is "outdated". The latest code is a bit different but probably similarly slow. I don't think I've made it any worse. We definitely need to look at updating the messages code in general, but I'm sort of waiting for a new use case to come along to justify a refactoring.

File: webapp/src/js/controllers/messages-content.js [hunk 16]
```diff
@@ -81,37 +81,23 @@ angular.module('inboxControllers').controller('MessagesContentCtrl',
...
-                                    .catch(err => {
-                                      if (err.code === 404) {
-                                        return Promise.resolve();
-                                      }
-
-                                      throw err;
-                                    });
+        return LineageModelGenerator.contact(id).catch(err => {
+          if (err.code === 404) {
+            return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/controllers/messages-content.js]: Done.

File: webapp/src/js/actions/messages.js (hunk 6 above)
Comment (garethbowen) [reply webapp/src/js/actions/messages.js]: Sold.

File: webapp/src/js/controllers/messages-content.js (hunk 7 above)
Comment (garethbowen) [reply webapp/src/js/controllers/messages-content.js]: Yeah. `markAllRead` meant all `messages` in the currently selected `conversation`. When you select a conversation from the LHS we assume you've read all the messages then displayed on the RHS.

File: webapp/src/js/controllers/messages.js (hunk 8 above)
Comment (garethbowen) [reply webapp/src/js/controllers/messages.js]: I don't think so but it's more readable so I've added them.

File: tests/e2e/send-message.js (hunk 9 above)
Comment (garethbowen) [reply tests/e2e/send-message.js]: Yeah... it's been a constant source of flakiness despite numerous attempts. This was the final straw for me  :(

**Comment by garethbowen** [comment]
Almost all suggestions incorporated. Another review, if you please!

File: webapp/src/js/controllers/messages.js (hunk 8 above)
Comment (kennsippell) [reply webapp/src/js/controllers/messages.js]: `false && false || true` evaluates to true
`false && (false || true) evalutes to false`

so i do think the parenthesis are required here

**Pull request merged**